File: app/src/main/java/kr/djspi/pipe01/fragment/DirectionDialog.java
package kr.djspi.pipe01.fragment;

import android.content.Context;
import android.content.DialogInterface;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.View.OnClickListener;
import android.view.ViewGroup;
import android.widget.FrameLayout;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.fragment.app.DialogFragment;
import androidx.fragment.app.FragmentManager;

import kr.djspi.pipe01.R;

import static android.view.View.GONE;
import static android.view.View.INVISIBLE;
import static android.view.View.VISIBLE;
import static kr.djspi.pipe01.BaseActivity.packageName;
import static kr.djspi.pipe01.BaseActivity.resources;
import static kr.djspi.pipe01.Const.PIPE_DIRECTIONS;
import static kr.djspi.pipe01.Const.TAG_DIRECTION;
import static kr.djspi.pipe01.Const.TAG_DISTANCE;
import static kr.djspi.pipe01.dto.PipeShape.PipeShapeEnum.parsePipeShape;
import static kr.djspi.pipe01.dto.SpiType.SpiTypeEnum.parseSpiType;

public class DirectionDialog extends DialogFragment implements OnClickListener {

    private static final String TAG = DirectionDialog.class.getSimpleName();
    private static final String DEF_TYPE = "drawable";
    private int selectIndex = -1;
    private int positionInt = -1;
    private String dialogTitle;
    private String typeString;
    private String shapeString;
    private String[] resIds;
    private Bundle bundle;
    private ImageView checkView;
    private FragmentManager fragmentManager;
    private OnSelectListener listener;

    public DirectionDialog() {
    }

    @Override
    public void onAttach(@NonNull Context context) {
        super.onAttach(context);
        if (context instanceof OnSelectListener) {
            listener = (OnSelectListener) context;
        }
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        fragmentManager = getFragmentManager();
        if (getArguments() != null) {
            bundle = getArguments();
            typeString = bundle.getString("typeString");
            shapeString = bundle.getString("shapeString");
            positionInt = bundle.getInt("positionInt");
        }
        dialogTitle = getString(R.string.popup_title_select_direction);
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_plot_direction, container, false);

        TextView titleView = view.findViewById(R.id.popup_title);
        titleView.setText(dialogTitle);

        view.findViewById(R.id.lay_2).setOnClickListener(this);
        view.findViewById(R.id.lay_4).setOnClickListener(this);
        view.findViewById(R.id.lay_6).setOnClickListener(this);
        view.findViewById(R.id.lay_8).setOnClickListener(this);

        checkView = view.findViewById(R.id.v_select);

        view.findViewById(R.id.button_close).setOnClickListener(this);
        view.findViewById(R.id.btn_cancel).setOnClickListener(this);
        view.findViewById(R.id.btn_ok).setOnClickListener(this);

        setLayoutVisibility(view);

        return view;
    }

    private void setLayoutVisibility(@NonNull View view) {
        ImageView image_2 = view.findViewById(R.id.image_2);
        ImageView image_8 = view.findViewById(R.id.image_8);
        ImageView image_4 = view.findViewById(R.id.image_4);
        ImageView image_6 = view.findViewById(R.id.image_6);

        resIds = new String[10];
        for (int i = 1; i <= 4; i++) {
            resIds[i * 2] = String.format("plan_%s_%s_%s_%string",
                    parseSpiType(typeString),
                    parsePipeShape(shapeString),
                    String.valueOf(positionInt),
                    PIPE_DIRECTIONS[i * 2]);
        }

        setImageView(image_2, resIds[2]);
        setImageView(image_4, resIds[4]);
        setImageView(image_6, resIds[6]);
        setImageView(image_8, resIds[8]);
    }

    private static void setImageView(ImageView view, String resId) {
        int i = resources.getIdentifier(resId, DEF_TYPE, packageName);
        if (i != 0) view.setBackgroundResource(i);
        else {
            FrameLayout layout = (FrameLayout) view.getParent();
            layout.setVisibility(GONE);
        }
    }

    @Override
    public void onClick(View v) {
        switch (v.getId()) {
            case R.id.btn_ok:
                if (selectIndex == -1) {
                    Toast.makeText(getContext(), "관로의 방향을 선택해주세요", Toast.LENGTH_LONG).show();
                    return;
                }
                listener.onSelect(TAG_DIRECTION, selectIndex, resIds[selectIndex]);
                if (positionInt == 5) {
                    dismissAllowingStateLoss();
                    return;
                } else {
                    DistanceDialog dialog = new DistanceDialog();
                    bundle.putString("planString", resIds[selectIndex]);
                    dialog.setArguments(bundle);
                    dialog.show(fragmentManager, TAG_DISTANCE);
                    dismissAllowingStateLoss();
                }
                break;
            case R.id.btn_cancel:
                listener.onSelect(TAG_DIRECTION, -2, (String) null);
                dismissAllowingStateLoss();
                break;
            case R.id.button_close:
                dismissAllowingStateLoss();
                break;
            case R.id.lay_2:
                selectIndex = 2;
                setFocus(v);
                break;
            case R.id.lay_8:
                selectIndex = 8;
                setFocus(v);
                break;
            case R.id.lay_4:
                selectIndex = 4;
                setFocus(v);
                break;
            case R.id.lay_6:
                selectIndex = 6;
                setFocus(v);
                break;
            default:
                break;
        }
    }

    private void setFocus(@NonNull View view) {
        checkView.setVisibility(INVISIBLE);
        view.findViewById(R.id.v_select).setVisibility(VISIBLE);
        this.checkView = view.findViewById(R.id.v_select);
    }

    @Override
    public void onDismiss(@NonNull DialogInterface dialog) {
        selectIndex = -1;
        resIds = null;
        super.onDismiss(dialog);
    }

    @Override
    public void onDetach() {
        super.onDetach();
        listener = null;
    }
}

File: app/src/main/java/kr/djspi/pipe01/fragment/DirectionDialog.kt
package kr.djspi.pipe01.fragment

import android.content.Context
import android.content.DialogInterface
import android.graphics.Color
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.View.*
import android.view.ViewGroup
import android.widget.FrameLayout
import android.widget.ImageView
import android.widget.TextView
import android.widget.Toast
import androidx.fragment.app.DialogFragment
import kr.djspi.pipe01.BaseActivity.Companion.defPackage
import kr.djspi.pipe01.Const.PIPE_DIRECTIONS
import kr.djspi.pipe01.Const.TAG_DIRECTION
import kr.djspi.pipe01.Const.TAG_DISTANCE
import kr.djspi.pipe01.R
import kr.djspi.pipe01.dto.PipeShape.PipeShapeEnum
import kr.djspi.pipe01.dto.PipeShape.PipeShapeEnum.Companion.parsePipeShape
import kr.djspi.pipe01.dto.SpiType.SpiTypeEnum.Companion.parseSpiType

class DirectionDialog : DialogFragment(), OnClickListener {

    private var selectIndex = -1
    private var positionInt = -1
    private var dialogTitle: String? = null
    private var typeString: String? = null
    private var shapeString: String? = null
    private var bundle: Bundle? = null
    private var resIds = arrayOfNulls<String>(10)
    private val selects = arrayOfNulls<ImageView>(10)
    private lateinit var dialogTitleSub: String
    private lateinit var listener: OnSelectListener

    override fun onAttach(context: Context) {
        super.onAttach(context)
        if (context is OnSelectListener) {
            listener = context
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        arguments?.let {
            bundle = it
            typeString = it.getString("typeString")
            shapeString = it.getString("shapeString")
            positionInt = it.getInt("positionInt")
        }
        dialogTitle = getString(R.string.popup_title_select_direction)
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        val view = inflater.inflate(R.layout.fragment_plot_direction, container, false)
        val title = view.findViewById<TextView>(R.id.popup_title)
        val titleSub = view.findViewById<TextView>(R.id.popup_title_sub)
        title.text = dialogTitle
        dialogTitleSub = when (shapeString) {
            PipeShapeEnum.엘보형135.type -> {
                titleSub.setTextColor(Color.parseColor("#F24E4E"))
                getString(R.string.popup_title_select_direction_sub_elb135)
            }
            else -> {
                titleSub.setTextColor(Color.BLACK)
                getString(R.string.popup_title_select_direction_sub)
            }
        }
        titleSub.text = dialogTitleSub
        arrayOf<View>(
            view.findViewById(R.id.lay_2),
            view.findViewById(R.id.lay_4),
            view.findViewById(R.id.lay_6),
            view.findViewById(R.id.lay_8),
            view.findViewById(R.id.button_close),
            view.findViewById(R.id.btn_cancel),
            view.findViewById(R.id.btn_ok)
        ).forEach {
            it.setOnClickListener(this)
        }
        setLayoutVisibility(view)
        return view
    }

    private fun setLayoutVisibility(view: View) {
        val defType = "id"
        for (i in 1..4) {
            resIds[i * 2] =
                "plan_${parseSpiType(typeString)}_${parsePipeShape(shapeString)}_${positionInt}_${PIPE_DIRECTIONS[i * 2]}"
            selects[i * 2] =
                view.findViewById(resources.getIdentifier("v_select_${i * 2}", defType, defPackage))
        }
        setImageView(view.findViewById(R.id.image_2), resIds[2]!!)
        setImageView(view.findViewById(R.id.image_4), resIds[4]!!)
        setImageView(view.findViewById(R.id.image_6), resIds[6]!!)
        setImageView(view.findViewById(R.id.image_8), resIds[8]!!)
    }

    private fun setImageView(view: ImageView, resId: String) {
        val i = resources.getIdentifier(resId, "drawable", defPackage)
        when {
            i != 0 -> view.setBackgroundResource(i)
            else -> {
                val layout = view.parent as FrameLayout
                layout.visibility = GONE
            }
        }
    }

    override fun onClick(v: View) {
        when (v.id) {
            R.id.btn_ok -> {
                if (selectIndex == -1) {
                    Toast.makeText(context, dialogTitleSub, Toast.LENGTH_LONG).show()
                    return
                }
                listener.onSelect(TAG_DIRECTION, selectIndex, resIds[selectIndex])
                if (positionInt == 5) {
                    dismissAllowingStateLoss()
                    return
                } else {
                    bundle!!.putString("planString", resIds[selectIndex])
                    DistanceDialog().apply { arguments = bundle }
                        .show(fragmentManager!!, TAG_DISTANCE)
                    dismissAllowingStateLoss()
                }
            }
            R.id.btn_cancel -> {
                listener.onSelect(TAG_DIRECTION, -2, null)
                dismissAllowingStateLoss()
            }
            R.id.button_close -> dismissAllowingStateLoss()
            R.id.lay_2 -> {
                selectIndex = 2
                setFocus(v, selectIndex)
            }
            R.id.lay_8 -> {
                selectIndex = 8
                setFocus(v, selectIndex)
            }
            R.id.lay_4 -> {
                selectIndex = 4
                setFocus(v, selectIndex)
            }
            R.id.lay_6 -> {
                selectIndex = 6
                setFocus(v, selectIndex)
            }
        }
    }

    private fun setFocus(view: View, selectIndex: Int) {
        selects.forEach {
            it?.visibility = INVISIBLE
        }
        selects[selectIndex]?.visibility = VISIBLE
    }

    override fun onDismiss(dialog: DialogInterface) {
        selectIndex = -1
        super.onDismiss(dialog)
    }
}
